### PR TITLE
CompilerMSL enhancements PLUS refactoring of MSL tests to avoid optimization crashes

### PR DESCRIPTION
**Pull request by billhollings**
This PR includes several MSL enhancements to support platforms and nested function arguments.

It also includes a refactoring of some of the MSL test shaders to run some of them without optimization. This is to partly address the requests in Issue #372.

SPIR-V optimizations is problematic for the reference shaders:

1. Optimization results change frequently...changing the reference shader output.

2. Many shaders cause the SPIR-V optimizer and validator to crash.

3. Some optimizations are not correct (eg. - `while(true)` instead of `for` loops).

As a temporary solution, to avoid slowing down SPIRV-Cross development, the problematic MSL shaders have been moved to the new `shaders-msl-no-opts` folder.

MSL shaders in the `shaders-msl` are run both with and without optimizations. MSL shaders in the `shaders-msl-no-opts` folder are only run WITHOUT SPIR-V optimizations.

There are similar problems with GLSL and HLSL shader optimizations...but I have not attempted to fix this. I expect the GitHub CI tests to fail on this PR as a result.

These optimization issues arose by via the latest SPIRV-Tools version.

I think a partial longer-term solution might include establishing SPIRV-Tools as a submodule of SPIRV-Cross...so that a given SPIRV-Cross commit would use a consistent SPIRV-Tools commit for optimization...ensuring that once working, the optimizations would remain unchanged.

### CODE REVIEW

**Comment by HansKristian-ARM** [comment]
For glslang/SPIRV-Tools, check checkout_glslang_spirv_tools.sh, this checks out a fixed version and builds it locally.

File: spirv_glsl.cpp [hunk 1]
```diff
@@ -5651,9 +5654,10 @@ void CompilerGLSL::emit_instruction(const Instruction &instruction)
 	case OpVectorTimesMatrix:
 	case OpMatrixTimesVector:
 	{
-		// If the matrix needs transpose, just flip the multiply order.
+		// If the matrix needs transpose and it is square, just flip the multiply order.
+		SPIRType *t;
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Should probably just assign t here, I feel it's a bit sketchky to assign in the branch.

File: spirv_glsl.hpp [hunk 2]
```diff
@@ -90,6 +90,13 @@ class CompilerGLSL : public Compiler
 		// If disabled on older targets, binding decorations will be stripped.
 		bool enable_420pack_extension = true;
 
+		// If true, all two-vector OpVectorShuffles will be forced to use a temporary variable.
```
Comment (HansKristian-ARM) [inline spirv_glsl.hpp]: This is a bit awkward to make into an option, this should just be forced.

File: spirv_msl.cpp [hunk 3]
```diff
@@ -2229,6 +2278,82 @@ string CompilerMSL::to_sampler_expression(uint32_t id)
 	return samp_id ? to_expression(samp_id) : to_expression(id) + sampler_name_suffix;
 }
 
+// Checks whether the ID is a row_major matrix that requires conversion before use
+bool CompilerMSL::is_non_native_row_major_matrix(uint32_t id)
+{
+	// Natively supported row-major matrices do not need to be converted.
+	// Legacy targets do not support row major.
+	if (backend.native_row_major_matrix && !is_legacy())
```
Comment (HansKristian-ARM) [inline spirv_msl.cpp]: is_legacy doesn't really make sense on MSL? it's more of a GLSL/HLSL thing.

File: spirv_msl.cpp [hunk 4]
```diff
@@ -2229,6 +2278,82 @@ string CompilerMSL::to_sampler_expression(uint32_t id)
...
+	const auto type = expression_type(id);
+	add_convert_row_major_matrix_function(type.columns, type.vecsize);
+	return true;
+}
+
+// Checks whether the member is a row_major matrix that requires conversion before use
+bool CompilerMSL::member_is_non_native_row_major_matrix(const SPIRType &type, uint32_t index)
+{
+	// Natively supported row-major matrices do not need to be converted.
+	if (backend.native_row_major_matrix && !is_legacy())
```
Comment (HansKristian-ARM) [inline spirv_msl.cpp]: is_legacy.

File: spirv_msl.cpp [hunk 5]
```diff
@@ -2599,7 +2722,7 @@ string CompilerMSL::entry_point_args(bool append_comma)
 			{
 				if (!ep_args.empty())
 					ep_args += ", ";
-				BuiltIn bi_type = meta[var_id].decoration.builtin_type;
+				BuiltIn bi_type = (BuiltIn)get_decoration(var_id, DecorationBuiltIn);
```
Comment (HansKristian-ARM) [inline spirv_msl.cpp]: This is a bit awkward with the casting back to an enum which meta[] does not need.

File: spirv_msl.cpp [hunk 6]
```diff
@@ -3298,9 +3424,44 @@ bool CompilerMSL::OpCodePreprocessor::handle(Op opcode, const uint32_t *args, ui
...
-	if (length > 1)
-		result_types[args[1]] = args[0];
+	// If it has one, keep track of the instruction's result type, mapped by ID
+	switch (opcode)
+	{
+	case OpStore:
+	case OpCopyMemory:
+	case OpCopyMemorySized:
+	case OpImageWrite:
+	case OpLoopMerge:
```
Comment (HansKristian-ARM) [inline spirv_msl.cpp]: Control flow related op-codes are unnecessary, like LoopMerge, SelectionMerge, Label, Branch*, Switch

File: spirv_msl.hpp [hunk 7]
```diff
@@ -77,10 +77,25 @@ class CompilerMSL : public CompilerGLSL
...
+		Platform platform = macOS;
 		uint32_t msl_version = make_msl_version(1, 2);
 		bool enable_point_size_builtin = true;
 		bool resolve_specialized_array_lengths = true;
 
+		bool is_ios()
+		{
+			return platform == iOS;
+		}
+		bool is_macos()
```
Comment (HansKristian-ARM) [inline spirv_msl.hpp]: Empty line between is_ios(), is_macos().

File: spirv_glsl.cpp (hunk 1 above)
Comment (billhollings) [reply spirv_glsl.cpp]: Done.

File: spirv_glsl.hpp (hunk 2 above)
Comment (billhollings) [reply spirv_glsl.hpp]: Always forcing makes sense for MSL. But maybe not GLSL and HLSL.

I've moved it to a backend config selection.

File: spirv_msl.cpp (hunk 3 above)
Comment (billhollings) [reply spirv_msl.cpp]: Copy and pasted code. Good catch. Removed now.

File: spirv_msl.cpp (hunk 4 above)
Comment (billhollings) [reply spirv_msl.cpp]: Copy and pasted code. Good catch. Removed now.

File: spirv_msl.cpp (hunk 5 above)
Comment (billhollings) [reply spirv_msl.cpp]: Reverted.

File: spirv_msl.hpp (hunk 7 above)
Comment (billhollings) [reply spirv_msl.hpp]: Done.

File: spirv_msl.cpp (hunk 5 above)
Comment (billhollings) [reply spirv_msl.cpp]: Removed them.

File: spirv_msl.cpp (hunk 6 above)
Comment (billhollings) [reply spirv_msl.cpp]: Removed them.

**Comment by billhollings** [comment]
I've update code as per the review. Thanks.

I'm now using the `checkout_glslang_spirv_tools.sh` script to lock in the correct `SPIRV-Tools` and `glslang` versions. That has fixed all but one of the shaders that were breaking under heavy `SPRIV-Tools` optimization.

I've left that one shader in `shaders-msl-no-opt`folder so that it is not run with optimizations. I will address why it is broken at a later date. It is a complicated nested function shader.

**Pull request merged**